Android 8+: Not allowed to start service Intent

**What problem do you get?**
I'm getting the exact same issue as ticket #1017 but when using `process.non-separate=true`.

**Which version of FileDownloader are you using when you produce such problem?**
I'm using the latest version: 1.7.4

**How to reproduce such problem?**
Add `process.non-separate=true` to filedownloader.properties.
Then follow steps as with ticket  #1017 

**Exception logcat** 
```
E: FATAL EXCEPTION: main
   Process: com.brightcove.player.test.exo2, PID: 9875
   java.lang.RuntimeException: Error receiving broadcast Intent { act=android.net.conn.CONNECTIVITY_CHANGE flg=0x4000010 (has extras) } in com.brightcove.player.network.ConnectivityMonitor$1@64fb744
       at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$-android_app_LoadedApk$ReceiverDispatcher$Args_52497(LoadedApk.java:1323)
       at android.app.-$Lambda$aS31cHIhRx41653CMnd4gZqshIQ.$m$7(Unknown Source:4)
       at android.app.-$Lambda$aS31cHIhRx41653CMnd4gZqshIQ.run(Unknown Source:39)
       at android.os.Handler.handleCallback(Handler.java:790)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:164)
       at android.app.ActivityThread.main(ActivityThread.java:6494)
       at java.lang.reflect.Method.invoke(Native Method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
    Caused by: java.lang.IllegalStateException: Not allowed to start service Intent { cmp=com.brightcove.player.test.exo2/com.liulishuo.filedownloader.services.FileDownloadService$SharedMainProcessService }: app is in background uid UidRecord{8d32f03 u0a355 LAST bg:+3m7s444ms idle change:cached procs:1 seq(0,0,0)}
       at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1521)
       at android.app.ContextImpl.startService(ContextImpl.java:1477)
       at android.content.ContextWrapper.startService(ContextWrapper.java:650)
       at com.liulishuo.filedownloader.FileDownloadServiceSharedTransmit.bindStartByContext(FileDownloadServiceSharedTransmit.java:145)
       at com.liulishuo.filedownloader.FileDownloadServiceProxy.bindStartByContext(FileDownloadServiceProxy.java:120)
       at com.liulishuo.filedownloader.FileDownloader.pauseAll(FileDownloader.java:282)
       at com.brightcove.player.network.DownloadManager$1.onConnectivityChanged(DownloadManager.java:120)
       at com.brightcove.player.network.ConnectivityMonitor$1.onReceive(ConnectivityMonitor.java:260)
       at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$-android_app_LoadedApk$ReceiverDispatcher$Args_52497(LoadedApk.java:1313)
       at android.app.-$Lambda$aS31cHIhRx41653CMnd4gZqshIQ.$m$7(Unknown Source:4) 
       at android.app.-$Lambda$aS31cHIhRx41653CMnd4gZqshIQ.run(Unknown Source:39) 
       at android.os.Handler.handleCallback(Handler.java:790) 
       at android.os.Handler.dispatchMessage(Handler.java:99) 
       at android.os.Looper.loop(Looper.java:164) 
       at android.app.ActivityThread.main(ActivityThread.java:6494) 
       at java.lang.reflect.Method.invoke(Native Method) 
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

--- a/library/src/main/java/com/liulishuo/filedownloader/FileDownloadServiceSharedTransmit.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/FileDownloadServiceSharedTransmit.java
@@ -25,6 +25,8 @@ import com.liulishuo.filedownloader.services.FDServiceSharedHandler;
 import com.liulishuo.filedownloader.services.FDServiceSharedHandler.FileDownloadServiceSharedConnection;
 import com.liulishuo.filedownloader.services.FileDownloadService.SharedMainProcessService;
 import com.liulishuo.filedownloader.util.DownloadServiceNotConnectedHelper;
+import com.liulishuo.filedownloader.util.FileDownloadLog;
+import com.liulishuo.filedownloader.util.FileDownloadUtils;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -142,7 +144,12 @@ class FileDownloadServiceSharedTransmit implements
             }
         }
         Intent i = new Intent(context, SERVICE_CLASS);
-        context.startService(i);
+        if (FileDownloadUtils.needMakeServiceForeground(context)) {
+            if (FileDownloadLog.NEED_LOG) FileDownloadLog.d(this, "start foreground service");
+            context.startForegroundService(i);
+        } else  {
+            context.startService(i);
+        }
     }
 
     @Override

--- a/library/src/main/java/com/liulishuo/filedownloader/download/CustomComponentHolder.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/download/CustomComponentHolder.java
@@ -17,10 +17,11 @@
 package com.liulishuo.filedownloader.download;
 
 import com.liulishuo.filedownloader.connection.FileDownloadConnection;
+import com.liulishuo.filedownloader.database.FileDownloadDatabase;
 import com.liulishuo.filedownloader.model.FileDownloadModel;
 import com.liulishuo.filedownloader.model.FileDownloadStatus;
 import com.liulishuo.filedownloader.services.DownloadMgrInitialParams;
-import com.liulishuo.filedownloader.database.FileDownloadDatabase;
+import com.liulishuo.filedownloader.services.ForegroundServiceConfig;
 import com.liulishuo.filedownloader.stream.FileDownloadOutputStream;
 import com.liulishuo.filedownloader.util.FileDownloadHelper;
 import com.liulishuo.filedownloader.util.FileDownloadLog;
@@ -41,6 +42,7 @@ public class CustomComponentHolder {
     private FileDownloadHelper.OutputStreamCreator outputStreamCreator;
     private FileDownloadDatabase database;
     private FileDownloadHelper.IdGenerator idGenerator;
+    private ForegroundServiceConfig foregroundServiceConfig;
 
     private static final class LazyLoader {
         private static final CustomComponentHolder INSTANCE = new CustomComponentHolder();
@@ -91,6 +93,19 @@ public class CustomComponentHolder {
         }
 
         return database;
+    }
+
+    public ForegroundServiceConfig getForegroundConfigInstance() {
+        if (foregroundServiceConfig != null) return foregroundServiceConfig;
+
+        synchronized (this) {
+            if (foregroundServiceConfig == null) {
+                foregroundServiceConfig = getDownloadMgrInitialParams()
+                        .createForegroundServiceConfig();
+            }
+        }
+
+        return foregroundServiceConfig;
     }
 
     public int getMaxNetworkThreadCount() {

--- a/library/src/main/java/com/liulishuo/filedownloader/services/BaseFileServiceUIGuard.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/services/BaseFileServiceUIGuard.java
@@ -155,7 +155,12 @@ public abstract class BaseFileServiceUIGuard<CALLBACK extends Binder, INTERFACE 
         }
 
         context.bindService(i, this, Context.BIND_AUTO_CREATE);
-        context.startService(i);
+        if (FileDownloadUtils.needMakeServiceForeground(context)) {
+            if (FileDownloadLog.NEED_LOG) FileDownloadLog.d(this, "start foreground service");
+            context.startForegroundService(i);
+        } else {
+            context.startService(i);
+        }
     }
 
     @Override
@@ -181,9 +186,13 @@ public abstract class BaseFileServiceUIGuard<CALLBACK extends Binder, INTERFACE 
     }
 
     public void startService(final Context context) {
-
         Intent i = new Intent(context, serviceClass);
-        context.startService(i);
+        if (FileDownloadUtils.needMakeServiceForeground(context)) {
+            if (FileDownloadLog.NEED_LOG) FileDownloadLog.d(this, "start foreground service");
+            context.startForegroundService(i);
+        } else {
+            context.startService(i);
+        }
     }
 
     protected abstract INTERFACE asInterface(IBinder service);

--- a/library/src/main/java/com/liulishuo/filedownloader/services/DownloadMgrInitialParams.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/services/DownloadMgrInitialParams.java
@@ -154,6 +154,27 @@ public class DownloadMgrInitialParams {
         }
     }
 
+    public ForegroundServiceConfig createForegroundServiceConfig() {
+        if (mMaker == null) {
+            return createDefaultForegroundServiceConfig();
+        }
+
+        final ForegroundServiceConfig foregroundServiceConfig = mMaker.mForegroundServiceConfig;
+        if (foregroundServiceConfig != null) {
+            if (FileDownloadLog.NEED_LOG) {
+                FileDownloadLog.d(this, "initial FileDownloader manager with the customize "
+                        + "foreground service config: %s", foregroundServiceConfig);
+            }
+            return foregroundServiceConfig;
+        } else {
+            return createDefaultForegroundServiceConfig();
+        }
+    }
+
+    private ForegroundServiceConfig createDefaultForegroundServiceConfig() {
+        return new ForegroundServiceConfig.Builder().needRecreateChannelId(true).build();
+    }
+
     private FileDownloadHelper.IdGenerator createDefaultIdGenerator() {
         return new DefaultIdGenerator();
     }
@@ -185,6 +206,7 @@ public class DownloadMgrInitialParams {
         FileDownloadHelper.ConnectionCreator mConnectionCreator;
         FileDownloadHelper.ConnectionCountAdapter mConnectionCountAdapter;
         FileDownloadHelper.IdGenerator mIdGenerator;
+        ForegroundServiceConfig mForegroundServiceConfig;
 
         /**
          * customize the id generator.
@@ -291,6 +313,15 @@ public class DownloadMgrInitialParams {
          */
         public InitCustomMaker connectionCreator(FileDownloadHelper.ConnectionCreator creator) {
             this.mConnectionCreator = creator;
+            return this;
+        }
+
+        /**
+         * customize configurations of foreground service
+         * @param config determines how to show an notification for the foreground service
+         */
+        public InitCustomMaker foregroundServiceConfig(ForegroundServiceConfig config) {
+            this.mForegroundServiceConfig = config;
             return this;
         }
 

--- a/library/src/main/java/com/liulishuo/filedownloader/services/FileDownloadService.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/services/FileDownloadService.java
@@ -17,11 +17,18 @@
 package com.liulishuo.filedownloader.services;
 
 import android.annotation.SuppressLint;
+import android.annotation.TargetApi;
+import android.app.NotificationChannel;
+import android.app.NotificationManager;
 import android.app.Service;
+import android.content.Context;
 import android.content.Intent;
+import android.os.Build;
 import android.os.IBinder;
 
+import com.liulishuo.filedownloader.download.CustomComponentHolder;
 import com.liulishuo.filedownloader.util.FileDownloadHelper;
+import com.liulishuo.filedownloader.util.FileDownloadLog;
 import com.liulishuo.filedownloader.util.FileDownloadProperties;
 import com.liulishuo.filedownloader.util.FileDownloadUtils;
 
@@ -65,12 +72,35 @@ public class FileDownloadService extends Service {
     @Override
     public int onStartCommand(Intent intent, int flags, int startId) {
         handler.onStartCommand(intent, flags, startId);
+        if (FileDownloadUtils.needMakeServiceForeground(this)) makeServiceForeground();
         return START_STICKY;
+    }
+
+    @TargetApi(Build.VERSION_CODES.O)
+    private void makeServiceForeground() {
+        ForegroundServiceConfig config =
+                CustomComponentHolder.getImpl().getForegroundConfigInstance();
+        if (FileDownloadLog.NEED_LOG) {
+            FileDownloadLog.d(this, "make service foreground: %s", config);
+        }
+        if (config.isNeedRecreateChannelId()) {
+            NotificationChannel notificationChannel = new NotificationChannel(
+                    config.getNotificationChannelId(),
+                    config.getNotificationChannelName(),
+                    NotificationManager.IMPORTANCE_LOW
+            );
+            NotificationManager notificationManager =
+                    (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+            if (notificationManager == null) return;
+            notificationManager.createNotificationChannel(notificationChannel);
+        }
+        startForeground(config.getNotificationId(), config.getNotification(this));
     }
 
     @Override
     public void onDestroy() {
         handler.onDestroy();
+        stopForeground(true);
         super.onDestroy();
     }
 

--- a/library/src/main/java/com/liulishuo/filedownloader/services/ForegroundServiceConfig.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/services/ForegroundServiceConfig.java
@@ -1,0 +1,153 @@
+/*
+ * Copyright (c) 2015 LingoChamp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.liulishuo.filedownloader.services;
+
+import android.annotation.TargetApi;
+import android.app.Notification;
+import android.content.Context;
+
+import com.liulishuo.filedownloader.R;
+import com.liulishuo.filedownloader.util.FileDownloadLog;
+
+@TargetApi(26)
+public class ForegroundServiceConfig {
+    private int notificationId;
+    private String notificationChannelId;
+    private String notificationChannelName;
+    private Notification notification;
+    private boolean needRecreateChannelId;
+
+    private ForegroundServiceConfig() {
+    }
+
+    private static final String DEFAULT_NOTIFICATION_CHANNEL_ID = "filedownloader_channel";
+    private static final String DEFAULT_NOTIFICATION_CHANNEL_NAME = "Filedownloader";
+    private static final int DEFAULT_NOTIFICATION_ID = android.R.drawable.arrow_down_float;
+
+    public int getNotificationId() {
+        return notificationId;
+    }
+
+    public String getNotificationChannelId() {
+        return notificationChannelId;
+    }
+
+    public String getNotificationChannelName() {
+        return notificationChannelName;
+    }
+
+    public Notification getNotification(Context context) {
+        if (notification == null) {
+            if (FileDownloadLog.NEED_LOG) {
+                FileDownloadLog.d(this, "build default notification");
+            }
+            notification = buildDefaultNotification(context);
+        }
+        return notification;
+    }
+
+    public boolean isNeedRecreateChannelId() {
+        return needRecreateChannelId;
+    }
+
+    public void setNotificationId(int notificationId) {
+        this.notificationId = notificationId;
+    }
+
+    public void setNotificationChannelId(String notificationChannelId) {
+        this.notificationChannelId = notificationChannelId;
+    }
+
+    public void setNotificationChannelName(String notificationChannelName) {
+        this.notificationChannelName = notificationChannelName;
+    }
+
+    public void setNotification(Notification notification) {
+        this.notification = notification;
+    }
+
+    public void setNeedRecreateChannelId(boolean needRecreateChannelId) {
+        this.needRecreateChannelId = needRecreateChannelId;
+    }
+
+    private Notification buildDefaultNotification(Context context) {
+        String title = context.getString(R.string.default_filedownloader_notification_title);
+        String content =
+                context.getString(R.string.default_filedownloader_notification_content);
+        Notification.Builder builder = new Notification.Builder(context, notificationChannelId);
+        builder.setContentTitle(title).setContentText(content)
+                .setSmallIcon(DEFAULT_NOTIFICATION_ID);
+        return builder.build();
+    }
+
+    @Override
+    public String toString() {
+        return "ForegroundServiceConfig{" +
+                "notificationId=" + notificationId +
+                ", notificationChannelId='" + notificationChannelId + '\'' +
+                ", notificationChannelName='" + notificationChannelName + '\'' +
+                ", notification=" + notification +
+                ", needRecreateChannelId=" + needRecreateChannelId +
+                '}';
+    }
+
+    public static class Builder {
+        private int notificationId;
+        private String notificationChannelId;
+        private String notificationChannelName;
+        private Notification notification;
+        private boolean needRecreateChannelId;
+
+        public Builder notificationId(int notificationId) {
+            this.notificationId = notificationId;
+            return this;
+        }
+
+        public Builder notificationChannelId(String notificationChannelId) {
+            this.notificationChannelId = notificationChannelId;
+            return this;
+        }
+
+        public Builder notificationChannelName(String notificationChannelName) {
+            this.notificationChannelName = notificationChannelName;
+            return this;
+        }
+
+        public Builder notification(Notification notification) {
+            this.notification = notification;
+            return this;
+        }
+
+        public Builder needRecreateChannelId(boolean needRecreateChannelId) {
+            this.needRecreateChannelId = needRecreateChannelId;
+            return this;
+        }
+
+        public ForegroundServiceConfig build() {
+            final ForegroundServiceConfig config = new ForegroundServiceConfig();
+            config.setNotificationChannelId(notificationChannelId == null ?
+                    DEFAULT_NOTIFICATION_CHANNEL_ID : notificationChannelId);
+            config.setNotificationChannelName(notificationChannelName == null ?
+                    DEFAULT_NOTIFICATION_CHANNEL_NAME : notificationChannelName);
+            config.setNotificationId(notificationId == 0 ?
+                    DEFAULT_NOTIFICATION_ID : notificationId);
+            config.setNeedRecreateChannelId(needRecreateChannelId);
+            config.setNotification(notification);
+            return config;
+        }
+    }
+}

--- a/library/src/main/java/com/liulishuo/filedownloader/services/ForegroundServiceConfig.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/services/ForegroundServiceConfig.java
@@ -96,13 +96,15 @@ public class ForegroundServiceConfig {
 
     @Override
     public String toString() {
-        return "ForegroundServiceConfig{" +
-                "notificationId=" + notificationId +
-                ", notificationChannelId='" + notificationChannelId + '\'' +
-                ", notificationChannelName='" + notificationChannelName + '\'' +
-                ", notification=" + notification +
-                ", needRecreateChannelId=" + needRecreateChannelId +
-                '}';
+        return "ForegroundServiceConfig{"
+                + "notificationId=" + notificationId
+                + ", notificationChannelId='" + notificationChannelId
+                + '\''
+                + ", notificationChannelName='" + notificationChannelName
+                + '\''
+                + ", notification=" + notification
+                + ", needRecreateChannelId=" + needRecreateChannelId
+                + '}';
     }
 
     public static class Builder {
@@ -139,12 +141,12 @@ public class ForegroundServiceConfig {
 
         public ForegroundServiceConfig build() {
             final ForegroundServiceConfig config = new ForegroundServiceConfig();
-            config.setNotificationChannelId(notificationChannelId == null ?
-                    DEFAULT_NOTIFICATION_CHANNEL_ID : notificationChannelId);
-            config.setNotificationChannelName(notificationChannelName == null ?
-                    DEFAULT_NOTIFICATION_CHANNEL_NAME : notificationChannelName);
-            config.setNotificationId(notificationId == 0 ?
-                    DEFAULT_NOTIFICATION_ID : notificationId);
+            config.setNotificationChannelId(notificationChannelId == null
+                    ? DEFAULT_NOTIFICATION_CHANNEL_ID : notificationChannelId);
+            config.setNotificationChannelName(notificationChannelName == null
+                    ? DEFAULT_NOTIFICATION_CHANNEL_NAME : notificationChannelName);
+            config.setNotificationId(notificationId == 0
+                    ? DEFAULT_NOTIFICATION_ID : notificationId);
             config.setNeedRecreateChannelId(needRecreateChannelId);
             config.setNotification(notification);
             return config;

--- a/library/src/main/java/com/liulishuo/filedownloader/util/FileDownloadUtils.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/util/FileDownloadUtils.java
@@ -848,8 +848,8 @@ public class FileDownloadUtils {
         String packageName = context.getApplicationContext().getPackageName();
         for (ActivityManager.RunningAppProcessInfo appProcess : appProcesses) {
             // The name of the process that this object is associated with.
-            if (appProcess.processName.equals(packageName) && appProcess.importance ==
-                    ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND) {
+            if (appProcess.processName.equals(packageName) && appProcess.importance
+                    == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND) {
                 return true;
             }
 

--- a/library/src/main/java/com/liulishuo/filedownloader/util/FileDownloadUtils.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/util/FileDownloadUtils.java
@@ -838,8 +838,11 @@ public class FileDownloadUtils {
         if (appProcesses == null) return false;
 
         PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
-        if (pm == null || !pm.isScreenOn()) {
-            return false;
+        if (pm == null) return false;
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT) {
+            if (!pm.isInteractive()) return false;
+        } else {
+            if (!pm.isScreenOn()) return false;
         }
 
         String packageName = context.getApplicationContext().getPackageName();

--- a/library/src/main/java/com/liulishuo/filedownloader/util/FileDownloadUtils.java
+++ b/library/src/main/java/com/liulishuo/filedownloader/util/FileDownloadUtils.java
@@ -23,6 +23,7 @@ import android.net.ConnectivityManager;
 import android.net.NetworkInfo;
 import android.os.Build;
 import android.os.Environment;
+import android.os.PowerManager;
 import android.os.StatFs;
 import android.text.TextUtils;
 
@@ -825,5 +826,35 @@ public class FileDownloadUtils {
 
     public static String defaultUserAgent() {
         return formatString("FileDownloader/%s", BuildConfig.VERSION_NAME);
+    }
+
+    private static boolean isAppOnForeground(Context context) {
+        ActivityManager activityManager = (ActivityManager) context.getApplicationContext()
+                .getSystemService(Context.ACTIVITY_SERVICE);
+        if (activityManager == null) return false;
+
+        List<ActivityManager.RunningAppProcessInfo> appProcesses =
+                activityManager.getRunningAppProcesses();
+        if (appProcesses == null) return false;
+
+        PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
+        if (pm == null || !pm.isScreenOn()) {
+            return false;
+        }
+
+        String packageName = context.getApplicationContext().getPackageName();
+        for (ActivityManager.RunningAppProcessInfo appProcess : appProcesses) {
+            // The name of the process that this object is associated with.
+            if (appProcess.processName.equals(packageName) && appProcess.importance ==
+                    ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND) {
+                return true;
+            }
+
+        }
+        return false;
+    }
+
+    public static boolean needMakeServiceForeground(Context context) {
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && !isAppOnForeground(context);
     }
 }

--- a/library/src/main/res/values/strings.xml
+++ b/library/src/main/res/values/strings.xml
@@ -1,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="default_filedownloader_notification_title">FileDownloader</string>
+    <string name="default_filedownloader_notification_content">FileDownloader is running.</string>
+</resources>

